src/api/apiErrors.js

```diff
@@ -0,0 +1,62 @@
+/* @flow strict-local */
+import type { ApiResponseErrorData } from './transportTypes';
+import * as logging from '../utils/logging';
+
+/**
+ * Runtime class of custom API error types.
+ */
+export class ApiError extends Error {
+  data: $ReadOnly<ApiResponseErrorData>;
+  httpStatus: number;
+
+  constructor(httpStatus: number, data: $ReadOnly<ApiResponseErrorData>) {
+    super(data.msg);
+    this.data = data;
+    this.httpStatus = httpStatus;
+  }
+}
+
+/**
+ * Given a server response (allegedly) denoting an error, produce an Error to be
+ * thrown.
+ *
+ * If the `data` argument is actually of the form expected from the server, the
+ * returned error will be an {@link ApiError}; otherwise it will be a generic
+ * Error.
+ */
+export const makeErrorFromApi = (httpStatus: number, data: ?mixed): Error => {
+  // Validate `data`, and construct the resultant error object.
+  if (data !== null && typeof data === 'object') {
+    if (typeof data.msg === 'string' && data.result === 'error') {
+      if ('code' in data && typeof data.code === 'string') {
+        /* $FlowFixMe, in the unlikely event we can ever narrow types with if
+        statements on properties. */
+        return new ApiError(httpStatus, data);
+      }
+    }
+  }
+
+  // Server has responded, but the response is not a valid error-object.
+  // (This should never happen, even on old versions of the Zulip server.)
+  logging.warn(`Bad response from server: ${JSON.stringify(data)}`);
+  return new Error('Server responded with invalid message');
+};
+
+/**
+ * The event queue is too old or has been garbage collected.
+ * https://zulipchat.com/api/get-events-from-queue#bad_event_queue_id-errors
+ */
+export const isEventQueueIdError = (e: Error): boolean =>
+  e instanceof ApiError && e.data.code === 'BAD_EVENT_QUEUE_ID';
```

Comment:
Like with `ApiError` itself, the commit adding this would be clarified a bit by squashing in the few-line change that switches to it. Much of the content of this is taken from the code deleted there, so that'd make the comparison a bit easier to look at.